### ISSUE TITLE
Cant fill target property if its readonly

### ISSUE THREAD

**Issue by Norbytus** [rekalogika/mapper#232]
I make transformer for Money BigDecimal

```php
<?php

declare(strict_types=1);

use Brick\Math\BigDecimal;
use Rekalogika\Mapper\Context\Context;
use Rekalogika\Mapper\Transformer\TransformerInterface;
use Rekalogika\Mapper\Transformer\TypeMapping;
use Rekalogika\Mapper\Util\TypeFactory;
use Symfony\Component\PropertyInfo\Type;

final readonly class BigDecimalTransformer implements TransformerInterface
{
    public function transform(
        mixed $source,
        mixed $target,
        ?Type $sourceType,
        ?Type $targetType,
        Context $context
    ): BigDecimal|string {
        if ($source instanceof BigDecimal) {
            return (string) $source;
        }

        return BigDecimal::of($source)->toBigDecimal();
    }

    public function getSupportedTransformation(): iterable
    {
        yield new TypeMapping(
            TypeFactory::string(),
            TypeFactory::objectOfClass(BigDecimal::class),
        );

        yield new TypeMapping(
            TypeFactory::int(),
            TypeFactory::objectOfClass(BigDecimal::class),
        );

        yield new TypeMapping(
            TypeFactory::float(),
            TypeFactory::objectOfClass(BigDecimal::class),
        );

        yield new TypeMapping(
            TypeFactory::objectOfClass(BigDecimal::class),
            TypeFactory::string(),
        );
    }
}
```

And two calasses

```php
final readonly class Source
{
    public function __construct(
        public string $amount,
    ) { }
}

final readonly class Target
{
    public function __construct(
        public BigDecimal $amount,
    ) { }
}
///With usage

$result = $this->mapper->map(new Source('10'), Target::class);

```

Get error 

```
In ReaderWriter.php line 217:

  Transformation of property "amount" on object type "Target" results in a different object instance from the original instance, but the new instance cannot be set on the target object. You
  may wish to add a setter method to the target class. Mapping path: "(root)".
  ```

But if delete ```readonly``` from target class it works find. I make same transformer for ```Brick\Money\Currency``` and its work find with readonly class

**Comment by priyadi** [comment]
Currently, Mapper does the mapping twice:

* On target instantiation. In your case, it maps the source `$amount` to `BigDecimal` `$amount` constructor argument.
* Mappings of the properties. In your case, it does the same thing again, but toward the public property `$amount` on the target side.

It fails on the second mapping because it doesn't have any way to set it on the target side (the class is read only).

I regard this as a bug, as most people would expect the second mapping should not take place. It should only take place if the target object `Target` already exists before the mapping.

In the meantime, you can drop the `readonly` like you mentioned.

Some Q and A for my record:

Why doesn't Mapper just skip the second mapping because the target is read only? Because we might be able to update the target object through wither methods or immutable setters. But we can't know in advance whether a method is a wither method or a fluent setter (can be both, too), other than from the name which can be unreliable.

If Mapper sees the transformer returns a different instance than the one already existing on the target side, it assumes it must be able to set the value back on the target host object. But it can't, so it throws an exception.

Why not just fail silently when that happens? If the target has wither methods, then the mappings will be lost. Better throw an exception than trying to debug why the values are not updated.

Why doesn't Mapper just skip the second mapping because it is already done for the constructor? It should, that should fix this particular case, but probably will not fix all the related edge cases.

**Comment by Norbytus** [comment]
But it's strange, and little bit confuse, because as i say with Currency it's work fine
```php
<?php

declare(strict_types=1);

use Brick\Money\Currency;
use Rekalogika\Mapper\Context\Context;
use Rekalogika\Mapper\Transformer\TransformerInterface;
use Rekalogika\Mapper\Transformer\TypeMapping;
use Rekalogika\Mapper\Util\TypeFactory;
use Symfony\Component\PropertyInfo\Type;

final readonly class CurrencyTransformer implements TransformerInterface
{
    public function transform(
        mixed $source,
        mixed $target,
        ?Type $sourceType,
        ?Type $targetType,
        Context $context,
    ): mixed {
        return Currency::of($source);
    }

    public function getSupportedTransformation(): iterable
    {
        yield new TypeMapping(
            TypeFactory::string(),
            TypeFactory::objectOfClass(Currency::class),
        );
    }
}

final readonly class Source
{
    public function __construct(public string $currency)
    {
        
    }
}

final readonly class Target
{
    public function __construct(public Currency $currency)
    {
        
    }
}
$result = $this->mapper->map(new Source('USD'), Target::class);
```

**Comment by priyadi** [comment]
The issue should be fixed by the linked PR, and in v1.12.0. The difference with BigDecimal & Currency is due to heuristics that Mapper uses to skip some mapping. I decided to remove the heuristics, and simply rely on skipping the mapping if it is already done on the constructor.
